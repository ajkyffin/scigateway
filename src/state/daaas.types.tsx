import { Step } from 'react-joyride';

export const AuthSuccessType = 'daaas:auth_success';
export const AuthFailureType = 'daaas:auth_failure';
export const LoadingAuthType = 'daaas:loading_auth';
export const LoadAuthProviderType = 'daaas:loading_auth_provider';
export const TokenExpiredType = 'daaas:expired_token';
export const ConfigureStringsType = 'daaas:configure_strings';
export const LoginType = 'daaas:login';
export const NotificationType = 'daaas:api:notification';
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';
export const SignOutType = 'daaas:signout';
export const ToggleDrawerType = 'daaas:toggledrawer';
export const DismissNotificationType = 'daaas:dismissnotification';
export const ConfigureFeatureSwitchesType = 'daaas:feature_switches';
export const AddNotificationsType = 'daaas:add_notification';
export const SiteLoadingType = 'daaas:site_loading';
export const ConfigureAnalyticsType = 'daaas:configure_analytics';
export const InitialiseAnalyticsType = 'daaas:initialise_analytics';
export const ToggleHelpType = 'daaas:toggle_help';
export const AddHelpTourStepsType = 'daaas:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
