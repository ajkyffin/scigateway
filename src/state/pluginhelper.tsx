import {
  GroupedPlugins,
  PluginConfig,
  RegisterRoutePayload,
} from './daaas.types';

// Custom comparer for plugin objects, sorts keys by order (asc), then by name (asc)
export const comparePlugins = (a: PluginConfig, b: PluginConfig): number =>
  a.order === b.order
    ? a.displayName.localeCompare(b.displayName)
    : a.order - b.order;

function buildMenuItemDictionary(plugins: PluginConfig[]): GroupedPlugins {
  const dict: GroupedPlugins = {};

  plugins.forEach(p => {
    if (!(p.section in dict)) {
      dict[p.section] = [];
    }
    dict[p.section].push(p);
  });

  return dict;
}

// Convert the list of plugins into a structured dataset for rendering
export function structureMenuData(plugins: PluginConfig[]): GroupedPlugins {
  const menuItems = buildMenuItemDictionary(plugins);

  Object.keys(menuItems).forEach(
    section => (menuItems[section] = menuItems[section].sort(comparePlugins))
  );
  return menuItems;
}

// Convert a plugin registration request body into a pluginConfig object
export function buildPluginConfig(payload: RegisterRoutePayload): PluginConfig {
  return { ...payload };
}
