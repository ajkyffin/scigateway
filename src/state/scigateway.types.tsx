import { Step } from 'react-joyride';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadAuthUrlType = 'scigateway:loading_auth_url';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
