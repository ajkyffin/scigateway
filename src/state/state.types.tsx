import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  avatarUrl: string;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  redirectUrl: string | null;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}
