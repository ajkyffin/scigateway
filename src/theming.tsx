import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export const buildTheme = (): Theme => {
  const options: UKRIThemeOptions = {
    typography: {
      useNextVariants: true,
    },
    palette: {
      primary: {
        main: '#1D4F91', // dark blue
      },
      secondary: {
        main: '#63666A', //grey
      },
    },
    ukri: {
      purple: '#8C4799',
      blue: '#1D4F91',
      orange: '#C34613',
      green: '#008275',
      grey: '#63666A',
    },
  };

  return createMuiTheme(options);
};
