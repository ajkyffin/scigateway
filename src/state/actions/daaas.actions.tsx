import { NotificationType, NotificationPayload } from '../daaas.types';
import { ActionType, StateType } from '../state.types';
import axios from 'axios';
import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';

export const daaasNotification = (
  message: string
): ActionType<NotificationPayload> => ({
  type: NotificationType,
  payload: {
    message,
  },
});

type MyThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export const configureSite = (): MyThunkResult<Promise<void>> => {
  return async dispatch => {
    const res = await axios.get(`/settings.json`);
    const settings = res.data;
    dispatch(daaasNotification(JSON.stringify(settings)));
  };
};
