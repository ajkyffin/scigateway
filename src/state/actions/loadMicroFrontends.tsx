/* eslint-disable @typescript-eslint/explicit-function-return-type */
// single-spa doesn't come with any types - all single-spa code should be limited to this file.
import * as singleSpa from 'single-spa';
import { Plugin } from '../state.types';

const runScript = async (url: string) => {
  return new Promise((resolve, reject) => {
    const script = document.createElement('script');
    script.src = url;
    script.onload = resolve;
    script.onerror = reject;

    const body = document.getElementsByTagName('body')[0];
    body.appendChild(script);
  });
};

const loadReactApp = async (name: string, url: string) => {
  await runScript(url);
  // Plugins are loaded on to the window and define their own property name so can't guess this at compile time
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  return (window as any)[name];
};

async function loadApp(name: string, appURL: string) {
  // register the app with singleSPA and pass a reference to the store of the app as well as a reference to the globalEventDistributor
  singleSpa.registerApplication(
    name,
    () => loadReactApp(name, appURL),
    (location: URL) => {
      return location.pathname === '/';
    }
  );
}

async function init(plugins: Plugin[]) {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const loadingPromises: Promise<any>[] = [];

  plugins
    .filter(p => p.enable)
    .forEach(p => {
      loadingPromises.push(loadApp(p.name, p.src));
    });

  // wait until all stores are loaded and all apps are registered with singleSpa
  await Promise.all(loadingPromises);

  singleSpa.start();
}

export default {
  init,
};
