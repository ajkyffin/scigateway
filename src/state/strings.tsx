import { AppStrings } from './scigateway.types';
import { StateType } from './state.types';

export function getAppStrings(
  state: StateType,
  section: string
): AppStrings | undefined {
  return state.scigateway.res ? state.scigateway.res[section] : undefined;
}

export const getString = (res: AppStrings | undefined, key: string): string =>
  (res && res[key]) || key;
