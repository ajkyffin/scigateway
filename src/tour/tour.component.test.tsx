import React from 'react';
import { act } from 'react-dom/test-utils';
import Tour from './tour.component';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import configureStore from 'redux-mock-store';
import { StateType } from '../state/state.types';
import { initialState } from '../state/reducers/scigateway.reducer';
import { createLocation } from 'history';
import { toggleHelp, toggleDrawer } from '../state/actions/scigateway.actions';
import { Provider } from 'react-redux';
import Joyride from 'react-joyride';
import { buildTheme } from '../theming';
import { MuiThemeProvider } from '@material-ui/core';
import TestAuthProvider from '../authentication/testAuthProvider';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class {
    public static placements = PopperJS.placements;

    public constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {},
      };
    }
  };
});

describe('Tour component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;
  const theme = buildTheme();

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();

    state = {
      scigateway: {
        ...initialState,
        helpSteps: [
          {
            target: '.test-1',
            content: 'Test 1',
          },
          {
            target: '.test-2',
            content: 'Test 2',
          },
          {
            target: '#plugin-link-test',
            content: 'Plugin link test',
          },
        ],
      },
      router: {
        action: 'POP',
        location: createLocation('/'),
      },
    };
    mockStore = configureStore();
  });

  it('renders correctly', () => {
    state.scigateway.showHelp = true;

    const wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <Tour store={mockStore(state)} />
      </MuiThemeProvider>
    );
    expect(
      wrapper
        .dive()
        .dive()
        .dive()
    ).toMatchSnapshot();
  });

  it('shows next tooltip when next is clicked', () => {
    state.scigateway.showHelp = true;
    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
            <div className="test-2" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    let joyride: Joyride = wrapper.find('Joyride').instance();
    expect(joyride.state.index).toEqual(0);

    wrapper
      .find('JoyrideTooltip')
      .find('button[aria-label="Next"]')
      .first()
      .simulate('click');

    expect(joyride.state.index).toEqual(1);
  });

  it('shows previous tooltip when back is clicked', () => {
    state.scigateway.showHelp = true;
    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
            <div className="test-2" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    let joyride: Joyride = wrapper.find('Joyride').instance();
    joyride.setState({ index: 1 });
    wrapper.update();
    expect(joyride.state.index).toEqual(1);

    wrapper
      .find('JoyrideTooltip')
      .find('button[aria-label="Back"]')
      .first()
      .simulate('click');

    expect(joyride.state.index).toEqual(0);
  });

  it('sends toggleHelp message when tour is finished', () => {
    state.scigateway.showHelp = true;
    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    wrapper
      .find('JoyrideTooltip')
      .find('button[aria-label="Close"]')
      .first()
      .simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleHelp());
  });

  it('sends toggleDrawer message when tour moves into plugin link tour steps', () => {
    state.scigateway.drawerOpen = false;
    state.scigateway.showHelp = true;
    state.scigateway.authorisation.provider = new TestAuthProvider(
      'test-token'
    );
    state.scigateway.helpSteps = [
      {
        target: '.test-1',
        content: 'Test 1',
      },
      {
        target: '#plugin-link-test',
        content: 'Plugin link test',
      },
    ];
    jest.useFakeTimers();

    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
            <div id="plugin-link-test" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    let joyride: Joyride = wrapper.find('Joyride').instance();
    expect(joyride.state.index).toEqual(0);

    wrapper
      .find('JoyrideTooltip')
      .find('button[aria-label="Next"]')
      .first()
      .simulate('click');

    act(() => {
      jest.runAllTimers();
    });

    expect(joyride.state.index).toEqual(1);

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });

  it('sends toggleDrawer message when tour moves out of plugin link tour steps', () => {
    state.scigateway.drawerOpen = true;
    state.scigateway.showHelp = true;
    state.scigateway.authorisation.provider = new TestAuthProvider(
      'test-token'
    );
    state.scigateway.helpSteps = [
      {
        target: '.test-1',
        content: 'Test 1',
      },
      {
        target: '#plugin-link-test',
        content: 'Plugin link test',
      },
    ];
    jest.useFakeTimers();

    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
            <div id="plugin-link-test" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    let joyride: Joyride = wrapper.find('Joyride').instance();
    joyride.setState({ index: 1 });
    wrapper.update();
    expect(joyride.state.index).toEqual(1);

    wrapper
      .find('JoyrideTooltip')
      .find('button[aria-label="Back"]')
      .first()
      .simulate('click');

    act(() => {
      jest.runAllTimers();
    });

    expect(joyride.state.index).toEqual(0);

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });

  it('does not show plugin links when user is not logged in', () => {
    state.scigateway.showHelp = true;
    state.scigateway.authorisation.provider = new TestAuthProvider(null);
    jest.useFakeTimers();

    const testStore = mockStore(state);

    const wrapper = mount(
      <MuiThemeProvider theme={theme}>
        <Provider store={testStore}>
          <div>
            <Tour />
            <div className="test-1" />
            <div className="test-2" />
            <div id="plugin-link-test" />
          </div>
        </Provider>
      </MuiThemeProvider>
    );

    let steps = wrapper.find('Joyride').prop('steps');
    expect(steps.length).toEqual(2);
    expect(steps).not.toContainEqual({
      target: '#plugin-link-test',
      content: 'Plugin link test',
      disableBeacon: true,
    });
  });
});
