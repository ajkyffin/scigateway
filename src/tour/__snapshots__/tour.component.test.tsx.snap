// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour component renders correctly 1`] = `
<div
  className="react-joyride"
>
  <JoyrideStep
    action="start"
    callback={[Function]}
    continuous={true}
    controlled={true}
    debug={false}
    helpers={
      Object {
        "close": [Function],
        "go": [Function],
        "info": [Function],
        "next": [Function],
        "open": [Function],
        "prev": [Function],
        "reset": [Function],
        "skip": [Function],
      }
    }
    index={0}
    lifecycle="init"
    setPopper={[Function]}
    shouldScroll={false}
    size={2}
    status="running"
    step={
      Object {
        "content": "Test 1",
        "disableBeacon": true,
        "disableCloseOnEsc": false,
        "disableOverlay": false,
        "disableOverlayClose": false,
        "disableScrollParentFix": false,
        "disableScrolling": false,
        "event": "click",
        "floaterProps": Object {
          "offset": 20,
          "options": Object {
            "preventOverflow": Object {
              "boundariesElement": "scrollParent",
            },
          },
          "styles": Object {
            "arrow": Object {
              "color": "#fff",
            },
            "options": Object {
              "zIndex": 1500,
            },
          },
          "wrapperOptions": Object {
            "offset": -18,
            "position": true,
          },
        },
        "hideBackButton": false,
        "locale": Object {
          "back": "Back",
          "close": "Close",
          "last": "Last",
          "next": "Next",
          "open": "Open the dialog",
          "skip": "Skip",
        },
        "offset": 10,
        "placement": "bottom",
        "showProgress": false,
        "showSkipButton": false,
        "spotlightClicks": false,
        "spotlightPadding": 10,
        "styles": Object {
          "beacon": Object {
            "WebkitAppearance": "none",
            "backgroundColor": "transparent",
            "border": 0,
            "borderRadius": 0,
            "color": "#555",
            "cursor": "pointer",
            "display": "inline-block",
            "fontSize": 16,
            "height": 36,
            "lineHeight": 1,
            "padding": 8,
            "position": "relative",
            "width": 36,
            "zIndex": 1500,
          },
          "beaconInner": Object {
            "animation": "joyride-beacon-inner 1.2s infinite ease-in-out",
            "backgroundColor": "#C34613",
            "borderRadius": "50%",
            "display": "block",
            "height": "50%",
            "left": "50%",
            "opacity": 0.7,
            "position": "absolute",
            "top": "50%",
            "transform": "translate(-50%, -50%)",
            "width": "50%",
          },
          "beaconOuter": Object {
            "animation": "joyride-beacon-outer 1.2s infinite ease-in-out",
            "backgroundColor": "rgba(195,70,19, 0.2)",
            "border": "2px solid #C34613",
            "borderRadius": "50%",
            "boxSizing": "border-box",
            "display": "block",
            "height": "100%",
            "left": 0,
            "opacity": 0.9,
            "position": "absolute",
            "top": 0,
            "transformOrigin": "center",
            "width": "100%",
          },
          "buttonBack": Object {
            "WebkitAppearance": "none",
            "backgroundColor": "transparent",
            "border": 0,
            "borderRadius": 0,
            "color": "#C34613",
            "cursor": "pointer",
            "fontSize": 16,
            "lineHeight": 1,
            "marginLeft": "auto",
            "marginRight": 5,
            "padding": 8,
          },
          "buttonClose": Object {
            "WebkitAppearance": "none",
            "backgroundColor": "transparent",
            "border": 0,
            "borderRadius": 0,
            "color": "#333",
            "cursor": "pointer",
            "fontSize": 16,
            "height": 14,
            "lineHeight": 1,
            "padding": 15,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 14,
          },
          "buttonNext": Object {
            "WebkitAppearance": "none",
            "backgroundColor": "#C34613",
            "border": 0,
            "borderRadius": 4,
            "color": "#fff",
            "cursor": "pointer",
            "fontSize": 16,
            "lineHeight": 1,
            "padding": 8,
          },
          "buttonSkip": Object {
            "WebkitAppearance": "none",
            "backgroundColor": "transparent",
            "border": 0,
            "borderRadius": 0,
            "color": "#333",
            "cursor": "pointer",
            "fontSize": 14,
            "lineHeight": 1,
            "padding": 8,
          },
          "options": Object {
            "arrowColor": "#fff",
            "backgroundColor": "#fff",
            "beaconSize": 36,
            "overlayColor": "rgba(0, 0, 0, 0.5)",
            "primaryColor": "#C34613",
            "spotlightShadow": "0 0 15px rgba(0, 0, 0, 0.5)",
            "textColor": "#333",
            "zIndex": 1500,
          },
          "overlay": Object {
            "backgroundColor": "rgba(0, 0, 0, 0.5)",
            "bottom": 0,
            "left": 0,
            "mixBlendMode": "hard-light",
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1500,
          },
          "overlayLegacy": Object {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1500,
          },
          "overlayLegacyCenter": Object {
            "backgroundColor": "rgba(0, 0, 0, 0.5)",
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1500,
          },
          "spotlight": Object {
            "backgroundColor": "gray",
            "borderRadius": 4,
            "position": "absolute",
          },
          "spotlightLegacy": Object {
            "borderRadius": 4,
            "boxShadow": "0 0 0 9999px rgba(0, 0, 0, 0.5), 0 0 15px rgba(0, 0, 0, 0.5)",
            "position": "absolute",
          },
          "tooltip": Object {
            "backgroundColor": "#fff",
            "borderRadius": 5,
            "boxSizing": "border-box",
            "color": "#333",
            "fontSize": 16,
            "maxWidth": "100%",
            "padding": 15,
            "position": "relative",
            "width": 380,
          },
          "tooltipContainer": Object {
            "lineHeight": 1.4,
            "textAlign": "center",
          },
          "tooltipContent": Object {
            "padding": "20px 10px",
          },
          "tooltipFooter": Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "flex-end",
            "marginTop": 15,
          },
          "tooltipFooterSpacer": Object {
            "flex": 1,
          },
          "tooltipTitle": Object {
            "fontSize": 18,
            "margin": 0,
          },
        },
        "target": ".test-1",
      }
    }
    update={[Function]}
  />
</div>
`;
